# Assignment 2
### Data Management & Analysis
### Oviya Adhan

The SQLite file "Edinburgh_Buses.db" describes the routes on the buses in the Edinburgh (Scotland) area at some time in the past. These routes were operated by quite a few different bus companies.

It contains two tables: "route" which contains the following fields:

1. "num" which is the route number (this sometimes also has some letters in it, like "46A")
2. "company" which is the bus company name
3. "pos" which is the position of the stop given by "stop" along the route
4. "stop" which is the stop number

and "stops" which contains:

1. "id" which is the stop number (joins with "stop" in the other table)
2. "stop name"

Note that two companies can use the same route number and those are different routes. For instance, in the data, companies "LRT" and "SMT" both operate a route 16.

The 11 questions on this assignment can be found below the startup code below. Please embed new code cells below each question. Please hand in a .ipynb file which can be exported from Colab. Each question is worth 3 points.

You may need to change the mount point of the SQLite file to your own Google Drive, and put the SQLite database file in your own Google Drive, in order to make the notebook work.


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# if you do not have write access to this path, you may need
# to change it to a path that you do have write access to.
gdrivePath="/content/drive/MyDrive/Data Science/Data Management & Analysis/"
import sqlite3
import pandas as pd
# create a connection to a database
conn = sqlite3.connect(gdrivePath+"Edinburgh_Buses.db")

In [ ]:
def execSQL(conn,query):
  conn.execute(query) # execute an SQL query
  conn.commit() # "commit" that query in order to make its action permanent

In [ ]:
def allrowsSelect(conn,query):
  cursor = conn.execute(query)
  for row in cursor:
   print(row)

In [ ]:
# sample query
import pandas as pd
pd.read_sql_query('select * from route where company="LRT"',conn)

num company  pos  stop
0     1     LRT    1   137
1     1     LRT    2    99
2     1     LRT    3    59
3     1     LRT    4    66
4     1     LRT    5    42
..   ..     ...  ...   ...
687   9     LRT    5    37
688   9     LRT    6    99
689   9     LRT    7   105
690   9     LRT    8   162
691   9     LRT    9   215

[692 rows x 4 columns]

1. Write a query to find all of the distinct bus companies.

In [ ]:
pd.read_sql_query('SELECT DISTINCT company FROM route', conn)

company
0      LRT
1      MAC
2      SMT
3      SMJ
4      CCC
5      MID
6      LOW
7      LRJ
8      WIL
9       FS
10      FF
11      JM
12     ROU
13      VL

2. Write a query that lists all of the distinct route number and company combinations.

In [ ]:
pd.read_sql_query('SELECT DISTINCT num, company FROM route', conn)

num company
0      1     LRT
1     10     LRT
2    100     LRT
3    100     MAC
4    106     SMT
..   ...     ...
127  D27     SMT
128  D28     SMT
129   R1     ROU
130   R2     ROU
131  VL2      VL

[132 rows x 2 columns]

3. Find a query to find the route numbers that are used by more than one company, and the number of times that they are used.

In [ ]:
#Note: On ED --> Professor said we need route number and number of distinct companies that use the route
pd.read_sql_query('SELECT num, COUNT(DISTINCT company) as qtyCompanies FROM route GROUP BY num HAVING qtyCompanies > 1', conn)

num  qtyCompanies
0   100             2
1    16             2
2    19             2
3    27             2
4    30             2
5    37             2
6    38             2
7    43             2
8    47             2
9    55             2
10   61             2
11   65             2
12   80             2
13   85             2
14   86             2
15   87             2

4. Write a query to find the names of all of the stops on route 17 operated by company LRT, ordered by position. Output the position, stop number, and stop name.

In [ ]:
pd.read_sql_query('SELECT route.pos, route.stop, stops.name FROM route LEFT JOIN stops ON route.stop = stops.id WHERE num = 17 AND company = "LRT" ORDER BY pos ASC', conn)

pos  stop            name
0    1   167        Newhaven
1    2   137           Leith
2    3   141      Leith Walk
3    4   194  Princes Street
4    5   230       Tollcross
5    6   160     Morningside
6    7   179         Oxgangs

5. Write a query that counts how many distinct routes each company has. It should output the company name and the number of routes.

In [ ]:
pd.read_sql_query('SELECT company, COUNT(DISTINCT num) as QtyRoutes FROM route GROUP BY company', conn)

company  QtyRoutes
0      CCC          1
1       FF          1
2       FS          2
3       JM          1
4      LOW          6
5      LRJ          1
6      LRT         72
7      MAC          1
8      MID          4
9      ROU          2
10     SMJ          1
11     SMT         38
12      VL          1
13     WIL          1

6. Write a query to find all of the distinct routes operated by company "LOW".

In [ ]:
pd.read_sql_query('SELECT DISTINCT num FROM route WHERE company = "LOW"', conn)

num
0   29
1   30
2  315
3   62
4   94
5   95

7. Find the ten routes with the most stops. Remember that a route is identified by both its company and number.

In [ ]:
pd.read_sql_query('SELECT num, company, COUNT(DISTINCT stop) as NumStops FROM route GROUP BY num ORDER BY NumStops DESC LIMIT 10', conn)

num company  NumStops
0  32     LRT        22
1  52     LRT        22
2  38     MID        21
3  19     SMT        18
4  30     LRT        18
5  43     SMT        18
6  80     SMT        18
7  16     SMT        17
8  47     MID        17
9  65     SMT        17

8. Find the ten routes with the fewest stops.

In [ ]:
pd.read_sql_query('SELECT num, company, COUNT(DISTINCT stop) as NumStops FROM route GROUP BY num ORDER BY NumStops LIMIT 10', conn)

num company  NumStops
0   60      JM         2
1   59     LRT         3
2   48     WIL         4
3  86A     LRT         4
4   93     SMT         4
5  VL2      VL         4
6   49     SMT         5
7   58     LRT         5
8   77     SMT         5
9   A8     SMT         5

9. Do any routes have the same stop on them more than once? Write a query to check if they do, by creating a table containing the num, company, and stop columns for all of the stops that appear more than one, along with the number of times that they appear.

In [ ]:
pd.read_sql_query('SELECT DISTINCT num, company, stop, COUNT(*) as appearances FROM route GROUP BY num, company, stop HAVING COUNT(stop) > 1', conn)

num company  stop  appearances
0     1     LRT   137            2
1    12     LRT    31            2
2    12     LRT   115            2
3    12     LRT   173            2
4    14     LRT   215            2
5    19     LRT    28            2
6   19A     LRT    28            2
7     2     LRT    31            2
8     2     LRT   115            2
9     2     LRT   173            2
10  275     SMT    48            2
11  275     SMT   165            2
12   2A     LRT    31            2
13   2A     LRT   115            2
14   2A     LRT   173            2
15   32     LRT   173            2
16   34     LRT    92            2
17   34     LRT   213            2
18   35     LRT    92            2
19   35     LRT   213            2
20   42     LRT    36            2
21  46A     LRT    36            2
22   52     LRT   173            2
23    6     LRT   137            2
24   81     LRT    47            2
25   81     LRT    59            2
26   81     LRT   162            2
27   81     LRT   219            2
28  81A     LRT    47            2
29  81A     LRT    59            2
30  81A     LRT   162            2
31  81A     LRT   219            2
32   R1     ROU   106            2
33   R2     ROU   106            2

10. Find the numbers and names of the ten stops that are on the most routes, along with the number of routes they are on. Count a stop only once for each route it appears on. For instance, stop 137 appears both at the beginning (position 1) and the end (position 11) of route 1 of company LRT; stop 137 should be counted only once for this route.

In [ ]:
pd.read_sql_query('SELECT DISTINCT route.stop, stops.name, COUNT(num) as NumRoutes FROM route LEFT JOIN stops ON route.stop = stops.id GROUP BY num ORDER BY NumRoutes DESC LIMIT 10', conn)

stop              name  NumRoutes
0   173           Niddrie         23
1   224          Stirling         21
2    44       Clovenstone         20
3   100        Gorebridge         20
4   219  St Andrew Square         19
5   154          Mayfield         19
6   173           Niddrie         18
7   208         Scotstoun         18
8   100        Gorebridge         18
9   137             Leith         18

11. Find all of the routes that go from stop 219 to stop 115. The position of stop 219 on the route needs to be before that of 115.

In [ ]:
pd.read_sql_query('SELECT A.num, A.company FROM route A, route B WHERE A.num = B.num AND A.company = B.company AND A.stop = 219 AND B.stop = 115 AND A.pos < B.pos ORDER BY A.num', conn)

num company
0  137     SMT
1   22     LRT
2  22A     LRT
3   25     LRT
4   37     SMT
5   38     MID
6   61     LRT
7   61     SMT
8   63     LRT
9  C12     SMT